File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/extractors/base/AbstractBookExtractor.java
package ru.kmorozov.gbd.core.logic.extractors.base;

import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import ru.kmorozov.db.core.config.IContextLoader;
import ru.kmorozov.db.core.logic.model.book.BookInfo;
import ru.kmorozov.gbd.core.logic.Proxy.HttpHostExt;
import ru.kmorozov.gbd.core.logic.connectors.Response;
import ru.kmorozov.gbd.core.logic.context.ContextProvider;
import ru.kmorozov.gbd.core.logic.context.ExecutionContext;
import ru.kmorozov.gbd.logger.Logger;
import ru.kmorozov.gbd.utils.HttpConnections;

import java.io.IOException;
import java.io.InputStream;
import java.net.UnknownHostException;
import java.nio.charset.Charset;

/**
 * Created by sbt-morozov-kv on 16.11.2016.
 */
public abstract class AbstractBookExtractor extends AbstractHttpProcessor {

    protected static final Logger logger = ExecutionContext.INSTANCE.getLogger(AbstractBookExtractor.class);

    protected String bookId;
    protected BookInfo bookInfo;

    public AbstractBookExtractor() {
    }

    protected AbstractBookExtractor(String bookId) {
        this(bookId, ContextProvider.getContextProvider());
    }

    protected AbstractBookExtractor(String bookId, final IContextLoader storedLoader) {
        this.bookId = bookId;

        BookInfo storedBookInfo = storedLoader == null ? null : storedLoader.getBookInfo(bookId);
        try {
            this.bookInfo = null == storedBookInfo ? this.findBookInfo() : storedBookInfo;
        } catch (final Exception e) {
            e.printStackTrace();

            this.bookInfo = null;
        }
    }

    protected abstract String getBookUrl();

    protected abstract String getReserveBookUrl();

    protected abstract BookInfo findBookInfo() throws Exception;

    protected Document getDocumentWithoutProxy() throws Exception {
        Connection.Response res = null;
        Document doc = null;

        try {
            res = Jsoup.connect(this.getBookUrl()).userAgent(HttpConnections.USER_AGENT).followRedirects(false).timeout(20000).method(Connection.Method.GET).execute();
        } catch (UnknownHostException uhe) {
            AbstractBookExtractor.logger.severe("Not connected to Internet!");
        } catch (Exception ex) {
            try {
                res = Jsoup.connect(this.getReserveBookUrl()).userAgent(HttpConnections.USER_AGENT).method(Connection.Method.GET).execute();
            } catch (Exception ex1) {
                throw new Exception(ex1);
            }
        }

        try {
            if (null != res) {
                doc = res.parse();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        return doc;
    }

    protected Document getDocumentWithProxy(HttpHostExt proxy) {
        Response resp = AbstractHttpProcessor.getContent(this.getBookUrl(), proxy, true);

        if (null == resp) return null;
        else {
            try (final InputStream is = resp.getContent()) {
                String respStr = new String(is.readAllBytes(), Charset.defaultCharset());
                return Jsoup.parse(respStr);
            } catch (IOException e) {
                return null;
            }
        }
    }

    public BookInfo getBookInfo() {
        return this.bookInfo;
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/extractors/base/AbstractBookExtractor.kt
package ru.kmorozov.gbd.core.logic.extractors.base

import org.jsoup.Connection
import org.jsoup.Connection.Method
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import ru.kmorozov.db.core.config.IContextLoader
import ru.kmorozov.db.core.logic.model.book.BookInfo
import ru.kmorozov.gbd.core.logic.Proxy.HttpHostExt
import ru.kmorozov.gbd.core.logic.context.ContextProvider
import ru.kmorozov.gbd.core.logic.context.ExecutionContext
import ru.kmorozov.gbd.utils.HttpConnections
import java.io.IOException
import java.net.UnknownHostException
import java.nio.charset.Charset

/**
 * Created by sbt-morozov-kv on 16.11.2016.
 */
abstract class AbstractBookExtractor : AbstractHttpProcessor {

    protected lateinit var bookId: String
    lateinit var bookInfo: BookInfo
        protected set

    protected abstract val bookUrl: String

    protected abstract val reserveBookUrl: String

    protected val documentWithoutProxy: Document?
        @Throws(Exception::class)
        get() {
            var res: Connection.Response? = null
            var doc: Document? = null

            try {
                res = Jsoup.connect(bookUrl).userAgent(HttpConnections.USER_AGENT).followRedirects(false).timeout(20000).method(Method.GET).execute()
            } catch (uhe: UnknownHostException) {
                logger.severe("Not connected to Internet!")
            } catch (ex: Exception) {
                try {
                    res = Jsoup.connect(reserveBookUrl).userAgent(HttpConnections.USER_AGENT).method(Method.GET).execute()
                } catch (ex1: Exception) {
                    throw Exception(ex1)
                }
            }

            try {
                if (null != res) {
                    doc = res.parse()
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }

            return doc
        }

    constructor() {}

    @JvmOverloads
    protected constructor(bookId: String, storedLoader: IContextLoader? = ContextProvider.getContextProvider()) {
        this.bookId = bookId

        val storedBookInfo = storedLoader?.getBookInfo(bookId)
        try {
            bookInfo = storedBookInfo ?: findBookInfo()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @Throws(Exception::class)
    protected abstract fun findBookInfo(): BookInfo

    protected fun getDocumentWithProxy(proxy: HttpHostExt): Document? {
        val resp = getContent(bookUrl, proxy, true)

        if (null == resp)
            return null
        else {
            try {
                resp.content.use { `is` ->
                    val respStr = String(`is`.readAllBytes(), Charset.defaultCharset())
                    return Jsoup.parse(respStr)
                }
            } catch (e: IOException) {
                return null
            }
        }
    }

    companion object {
        protected val logger = ExecutionContext.INSTANCE.getLogger(AbstractBookExtractor::class.java)
    }
}
